nixos/mattermost: Clean up, modernize, support MySQL, support mmctl, and support plugin building

numinit opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

nixpkgs deserves a Mattermost module that uses more up to date best practices! 

Taking ideas from #198040, we should support more complex database configurations and better defaults like peer authentication.

Enable mmctl and the Mattermost test suite (~17000 tests).

Support building Mattermost plugins in the sandbox (couple examples are in the test suite).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

numinit commented:
See the tests for an example of how to add and enable a plugin.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1619

Kranzes commented:
To have a truly immutable plugins install & enable functionality we will need to patch Mattermost's source code. But for now this is fine I guess...

SuperSandro2000 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -166,179 +207,126 @@ in
+    systemd.services.mattermost = {
+      description = "Mattermost chat service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" "postgresql.service" ];
+      requires = [ "network.target" "postgresql.service" ];
```
There should be an option to run the postgres on another machine or not use it at all.

SuperSandro2000 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -94,49 +85,99 @@ let
-        default = "Mattermost";
-        description = lib.mdDoc "Name of this Mattermost site.";
+        example = "https://chat.example.com";
+        description = lib.mdDoc "URL the Mattermost instance is reachable under, without trailing slash.";
```
This should be asserted

SuperSandro2000 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -94,49 +85,99 @@ let
-
 {
+  imports = [
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "enable" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "package" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "parameters" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
```
We should move those to its own module instead.

SuperSandro2000 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -94,49 +85,99 @@ let
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "enable" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "package" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "parameters" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "localDatabaseCreate" ] "The PostgreSQL database now gets created automatically by default.")
```
I didn't see an option to turn this off

SuperSandro2000 commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -94,49 +85,99 @@ let
+    (mkRemovedOptionModule [ "services" "mattermost" "matterircd" "parameters" ] "This option has been removed as it shouldn't be part of the Mattermost module.")
+    (mkRemovedOptionModule [ "services" "mattermost" "localDatabaseCreate" ] "The PostgreSQL database now gets created automatically by default.")
+    (mkRemovedOptionModule [ "services" "mattermost" "localDatabasePassword" ] "The database is no longer secured via password but rather with UNIX user authentication.")
+    (mkRemovedOptionModule [ "services" "mattermost" "localDatabaseUser" ] "The database user now matches the user under which Mattermost runs under.")
```
This does not always work and have an option.

numinit commented:
Haven't forgotten about this. I think the strategy will be to use the current Mattermost module as a reference to re-add the missing features.

numinit committed:
nixos/mattermost: modernize, support MySQL and mmctl

Based on #198040. Prioritizes backwards compatibility, including
database and plugin compatibility, while adding more sensible
defaults like database peer authentication.

Expand the scope of tests to include plugins (including building
from source) and testing that a piece of media uploads and downloads
to make sure the storage directory doesn't vanish.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -1,32 +1,87 @@
+      if value == null then null else (escapeURL name) + "=" + (escapeURL (toString value))
+    );
+
+    userPart =
+      if user == null && password == null then
+        ""
+      else if user != null && password != null then
+        escapeURL user + ":" + escapeURL password
+      else
+        escapeURL (if password != null then password else user);
```
Technically a username (without a password) is valid. I'm interpreting a password without a username as a sort of API key usecase, though this is more for URIs in general even though not necessarily Postgres ones.

Jaculabilis replied:
In a URL with `apikey@host`, `apikey` is parsed as a username, even if its function is closer to a password. If someone wants to connect to their postgres with an API key, they probably know enough about URLs to know they should use the username field.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -261,73 +428,148 @@ in
 
   config = mkMerge [
     (mkIf cfg.enable {
-      users.users = optionalAttrs (cfg.user == "mattermost") {
-        mattermost = {
-          group = cfg.group;
-          uid = config.ids.uids.mattermost;
-          home = cfg.statePath;
-        };
+      users.users.${cfg.user} = {
+        group = cfg.group;
+        uid = config.ids.uids.mattermost;
+        home = cfg.dataDir;
```
Thanks, I think there was a reason this was added from the blame and I'll change it back.

numinit replied:
I think I figured it out. The secret is to use `uid = mkIf (cfg.user == "mattermost") config.ids.uids.mattermost;` since then we auto-assign a UID if and only if the username is nonstandard. Same with gid. Without this, my tests with a nonstandard username failed because there was no user created.

Jaculabilis replied:
I don't know how many NixOS Mattermost deployments there are or how many of them are changing the user, but suppose someone was running it as their own personal user. With the previous behavior, no user configuration is done. With the current behavior, this would cause a conflict on `group` and `home` requiring `mkForce` or similar.

Maybe there needs to be a release notes item to the effect of "if you run mattermost with a nonstandard user, you should watch out/review the changes"? I still have no idea whether we're breaking anyone with this or just fixing a potential bug.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -261,73 +428,148 @@ in
+        "L+ ${cfg.dataDir}/templates - - - - ${cfg.package}/templates"
+        "L+ ${cfg.dataDir}/client - - - - ${cfg.package}/client"
+      ] ++ (
+        if mattermostPlugins == null then
+          # Create the plugin tarball directory if it's a symlink.
+          [
+            "r ${cfg.dataDir}/plugins - - - - -"
+            "d ${cfg.dataDir}/plugins 0750 ${cfg.user} ${cfg.group} - -"
+          ]
+        else
+          # Symlink the plugin tarball directory.
+          ["L+ ${cfg.dataDir}/plugins - - - - ${mattermostPlugins}"]
+      );
```
The directory gets removed because that's what `L+` does.

numinit commented on nixos/modules/services/web-apps/mattermost.nix:
```diff
@@ -261,73 +428,148 @@ in
+        "r ${pluginDir} - - - - -"
+        "d ${pluginDir} 0750 ${cfg.user} ${cfg.group} - -"
+        "d ${mattermostConf.PluginSettings.Directory} 0750 ${cfg.user} ${cfg.group} - -"
+        "d ${mattermostConf.PluginSettings.ClientDirectory} 0750 ${cfg.user} ${cfg.group} - -"
+
+        "L+ ${cfg.dataDir}/fonts - - - - ${cfg.package}/fonts"
```
Done

numinit committed:
mattermost: enable test suite, mmctl support, and plugin building

Add the following passthrus:

- mattermost.withTests (builds with tests but takes a while)
- mattermost.withoutTests (builds without tests but is shorter)
- mattermost.tests.mattermostWithTests
- mattermost.buildPlugin (builds a plugin from source)

Enable the mmctl build, and its tests, if selected.

numinit commented:
@ofborg eval mattermost

numinit commented:
Still a draft, remaining tasks are:

- MySQL config support, since we've got it working in the checkPhase tests now
- Get the NixOS smoke test fully working, including with MySQL
- buildMattermostPlugin function

numinit commented:
@ofborg eval mattermost

numinit commented:
@ofborg eval mattermost

numinit commented on pkgs/by-name/ma/mattermost/package.nix:
```diff
@@ -1,88 +1,135 @@
-  # Makes nix-update-script pick up the fetchurl for the webapp.
-  # https://github.com/Mic92/nix-update/blob/1.3.1/nix_update/eval.py#L179
-  offlineCache = webapp;
+    webapp = fetchurl {
```
Coming for this next...

numinit replied:
Sadly does not appear to be possible since the lockfile is inconsistent with hashes.

Jaculabilis commented on nixos/tests/mattermost/test.jpg:
incredible, 10/10, no notes

numinit commented:
@ofborg eval mattermost

numinit committed:
release-notes/25.05: add Mattermost details

numinit commented:
@ofborg eval mattermost

numinit commented:
@ofborg eval mattermost

numinit commented:
@ofborg eval mattermost.withTests

numinit committed:
mmctl: use correct Mattermost derivation

We don't have to run the full Mattermost/mmctl e2e test suite for this
derivation.

numinit commented:
Drafting (again) because of ARM issues.

numinit commented:
ARM issues resolved, there was one more broken test suite that was also broken on x86. Ready to go!

numinit commented:
@ofborg eval mattermost

numinit committed:
manual: add mattermost.chapter.md

fricklerhandwerk commented on nixos/tests/mattermost/test.jpg:
:rofl:

fricklerhandwerk reviewed (approved):
Wow, now that's dedication. Thanks a lot for this!

Great docs btw

numinit commented:
Drafting (again): as discussed in https://github.com/NixOS/nixpkgs/pull/370450#issuecomment-2568620553 we'd like a "latest" build of Mattermost to land first so update script changes can easily be backported, then this can land.

numinit commented:
Progress update on this: I've locally rebased against #373113, and made the tests take both ESR and latest into account. Once that PR closes, I'll rebase again and undraft, because it's looking like it's time.

I'm happy to report that all {unit,integration} tests on {patched,unpatched} webapps for {mutable,mostly mutable,immutable,environment file} configs work for Mattermost {ESR,latest} on {MySQL,Postgres}, including plugin builds and migrations from ESR to latest. Have to test on ARM64 a little more, but it's pretty promising. The tests that we have to disable are fairly consistently the ones that use SMTP and Elasticsearch, plus a handful that seem plain broken.

numinit committed:
{mattermost,mattermostLatest}: fix string escapes

numinit commented:
Going to make sure there are no more ARM issues and then undraft.

numinit commented:
We're good. All builds and tests, including NixOS tests, pass on both my Zen 4 EPYC box and the aarch64 community builder.

numinit commented:
:confetti_ball: 2 years in the making, thanks for the merge!

fricklerhandwerk commented:
@numinit thanks a lot for sticking at it. Contributors like you are the true heroes of NixOS.

numinit commented:
Very kind words @fricklerhandwerk  - thank YOU for your tireless work on all things Nix too!
